src/runtime/security/negotiation_manager.cpp

```diff
@@ -67,13 +69,33 @@ void negotiation_service::on_negotiation_request(negotiation_rpc rpc)
             static_cast<server_negotiation *>(_negotiations[rpc.dsn_request()->io_session].get());
     }
 
-    dassert(srv_negotiation != nullptr,
-            "negotiation is null for msg: {}",
-            rpc.dsn_request()->rpc_code().to_string());
+    if (nullptr == srv_negotiation) {
+        derror_f("negotiation is null for msg: {}", rpc.dsn_request()->rpc_code().to_string());
+        return;
+    }
     srv_negotiation->handle_request(rpc);
 }
 
-void negotiation_service::on_rpc_connected(rpc_session *session)
+void negotiation_manager::on_negotiation_response(error_code err, negotiation_rpc rpc)
+{
+    dassert(rpc.dsn_request()->io_session->is_client(),
+            "only client session receives negotiation response");
+
+    client_negotiation *cli_negotiation = nullptr;
+    {
+        utils::auto_read_lock l(_lock);
+        cli_negotiation =
+            static_cast<client_negotiation *>(_negotiations[rpc.dsn_request()->io_session].get());
+    }
+
+    if (nullptr == cli_negotiation) {
+        derror_f("negotiation is null for msg: {}", rpc.dsn_request()->rpc_code().to_string());
```

Comment:
What do we gonna do for such a message? Isn't there any error code? Any response? Does that mean the client will keep waiting but the server just ignores it?

Reply:
in this case, the session is closed already. so we don't need send a response here.

Reply:
I can't see why it's closed? If it's closed, why it wasn't removed from `_negotiations`?

Reply:
If `client_negotiation` sends a message to `server_negotiation`,  and the `server_negotiation` find something is wrong when it receives this message, it will close the connection. 
On the client side, it will remove the negotiation from `_negotiation`.
And then the message which is send by client_negotiation  is timeout. It will calls the corresponding callback, which is `on_negotiation_response`. And in this function, it will not find the negotiation, because it is already removed. So we should add a judgement here.

Reply:
```
    {
        utils::auto_read_lock l(_lock);
        cli_negotiation =
            static_cast<client_negotiation *>(_negotiations[rpc.dsn_request()->io_session].get());
    }
    if (nullptr == cli_negotiation) {
        derror_f("negotiation is null for msg: {}", rpc.dsn_request()->rpc_code().to_string());
```

This is not proper usage of std::map. Actually `_negotiations[rpc.dsn_request()->io_session]` is not a read-only operation. It WRITES to the map. I think you misunderstood this, a very common mistake that would be made for new c++er. This code will incur thread-conflict as the section was wrapped by read-lock only, rather than write-lock. 

If you want to check if the `client_negotiation` was removed, use `std::map::find`.

```
    {
        utils::auto_read_lock l(_lock);
        auto it = _negotiations.find(rpc.dsn_request()->io_session);
        if(it == _negotiations.end()) {
            return; // ignore
        }
        cli_negotiation = static_cast<client_negotiation *>(it->second.get())
    }
    cli_negotiation->handle_response(err, std::move(rpc.response()));
```